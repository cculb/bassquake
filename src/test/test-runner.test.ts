import { describe, it, expect, beforeEach } from 'vitest'

describe('Test Runner Validation', () => {
  describe('Environment Setup', () => {
    it('has access to vitest globals', () => {
      expect(describe).toBeDefined()
      expect(it).toBeDefined()
      expect(expect).toBeDefined()
    })

    it('has jsdom environment configured', () => {
      expect(document).toBeDefined()
      expect(window).toBeDefined()
      expect(global).toBeDefined()
    })

    it('has proper TypeScript support', () => {
      // TypeScript compilation should work
      const testString: string = 'test'
      const testNumber: number = 42
      const testBoolean: boolean = true
      
      expect(typeof testString).toBe('string')
      expect(typeof testNumber).toBe('number')
      expect(typeof testBoolean).toBe('boolean')
    })
  })

  describe('Mock Functionality', () => {
    it('can create and use mocks', async () => {
      const { vi } = await import('vitest')
      const mockFn = vi.fn()
      
      mockFn('test')
      expect(mockFn).toHaveBeenCalledWith('test')
      expect(mockFn).toHaveBeenCalledTimes(1)
    })

    it('has audio context mocks available', () => {
      expect(AudioContext).toBeDefined()
      expect(typeof AudioContext).toBe('function')
      
      const audioContext = new AudioContext()
      expect(audioContext.createGain).toBeDefined()
      expect(audioContext.createOscillator).toBeDefined()
    })

    it('has media devices mocks available', () => {
      expect(navigator.mediaDevices).toBeDefined()
      expect(navigator.mediaDevices.getUserMedia).toBeDefined()
    })
  })

  describe('Testing Library Integration', () => {
    it('has react testing library available', async () => {
      const { render, screen } = await import('@testing-library/react')
      expect(render).toBeDefined()
      expect(screen).toBeDefined()
    })

    it('has user event utilities available', async () => {
      const userEvent = await import('@testing-library/user-event')
      expect(userEvent.default.setup).toBeDefined()
    })

    it('has jest-dom matchers available', () => {
      // These matchers should be available due to setup file
      expect(expect.extend).toBeDefined()
    })
  })

  describe('Performance Validation', () => {
    it('has performance API available', () => {
      expect(performance).toBeDefined()
      expect(performance.now).toBeDefined()
      expect(typeof performance.now()).toBe('number')
    })

    it('has requestAnimationFrame mock', () => {
      expect(requestAnimationFrame).toBeDefined()
      expect(cancelAnimationFrame).toBeDefined()
    })

    it('can measure test execution time', () => {
      const start = performance.now()
      
      // Simulate some work
      let sum = 0
      for (let i = 0; i < 1000; i++) {
        sum += i
      }
      
      const end = performance.now()
      const duration = end - start
      
      expect(duration).toBeGreaterThan(0)
      expect(duration).toBeLessThan(100) // Should be very fast
      expect(sum).toBe(499500) // Verify calculation worked
    })
  })

  describe('Error Handling', () => {
    it('can catch and test errors', () => {
      const throwError = () => {
        throw new Error('Test error')
      }

      expect(throwError).toThrow('Test error')
      expect(throwError).toThrowError(Error)
    })

    it('can test async error handling', async () => {
      const asyncError = async () => {
        throw new Error('Async error')
      }

      await expect(asyncError()).rejects.toThrow('Async error')
    })

    it('can test promise rejections', async () => {
      const rejectedPromise = Promise.reject(new Error('Promise rejection'))
      
      await expect(rejectedPromise).rejects.toThrow('Promise rejection')
    })
  })

  describe('Async Testing Support', () => {
    it('supports async/await in tests', async () => {
      const asyncFunction = async () => {
        return new Promise((resolve) => {
          setTimeout(() => resolve('async result'), 10)
        })
      }

      const result = await asyncFunction()
      expect(result).toBe('async result')
    })

    it('supports promise-based testing', () => {
      const promiseFunction = () => Promise.resolve('promise result')
      
      return expect(promiseFunction()).resolves.toBe('promise result')
    })

    it('handles timeouts correctly', async () => {
      const slowFunction = () => {
        return new Promise((resolve) => {
          setTimeout(() => resolve('slow result'), 50)
        })
      }

      const result = await slowFunction()
      expect(result).toBe('slow result')
    }, 1000) // 1 second timeout
  })

  describe('Mock Cleanup', () => {
    it('resets mocks between tests', async () => {
      const { vi } = await import('vitest')
      const mockFn = vi.fn()
      
      // This should start fresh even if other tests used mocks
      expect(mockFn).not.toHaveBeenCalled()
      
      mockFn('test')
      expect(mockFn).toHaveBeenCalledTimes(1)
    })

    it('has clean global state', () => {
      // Document should be reset between tests
      expect(document.body.innerHTML).toBe('')
      
      // No leftover event listeners from previous tests
      expect(window.onload).toBeNull()
    })
  })

  describe('Coverage Support', () => {
    it('tracks function coverage', () => {
      const testFunction = (x: number) => x * 2
      
      // This function call should be tracked by coverage
      const result = testFunction(21)
      expect(result).toBe(42)
    })

    it('tracks branch coverage', () => {
      const conditionalFunction = (condition: boolean) => {
        if (condition) {
          return 'true branch'
        } else {
          return 'false branch'
        }
      }

      // Both branches should be covered
      expect(conditionalFunction(true)).toBe('true branch')
      expect(conditionalFunction(false)).toBe('false branch')
    })

    it('tracks line coverage', () => {
      let counter = 0
      counter++ // This line should be covered
      counter += 10 // This line should be covered
      counter *= 2 // This line should be covered
      
      expect(counter).toBe(22)
    })
  })

  describe('Test Isolation', () => {
    let testVariable = 0

    beforeEach(() => {
      testVariable = 0
    })

    it('maintains test isolation - test 1', () => {
      testVariable = 1
      expect(testVariable).toBe(1)
    })

    it('maintains test isolation - test 2', () => {
      // This should start with the original value, not affected by test 1
      expect(testVariable).toBe(0)
      testVariable = 2
      expect(testVariable).toBe(2)
    })

    it('maintains test isolation - test 3', () => {
      // This should also start with the original value
      expect(testVariable).toBe(0)
    })
  })

  describe('Snapshot Testing Support', () => {
    it('can create and match snapshots', () => {
      const testObject = {
        name: 'Bassquake',
        version: '1.0.0',
        features: ['sequencer', 'synthesizer', 'motion-control']
      }

      expect(testObject).toMatchSnapshot()
    })

    it('handles inline snapshots', () => {
      const simpleValue = 'test-value'
      expect(simpleValue).toMatchInlineSnapshot('"test-value"')
    })
  })

  describe('Test Suite Performance', () => {
    it('completes individual tests quickly', () => {
      const start = performance.now()
      
      // Simulate typical test operations
      const data = Array(1000).fill(0).map((_, i) => i)
      const sum = data.reduce((acc, val) => acc + val, 0)
      
      const end = performance.now()
      const duration = end - start
      
      expect(sum).toBe(499500)
      expect(duration).toBeLessThan(50) // Should complete in < 50ms
    })

    it('handles multiple assertions efficiently', () => {
      const start = performance.now()
      
      // Multiple assertions should be fast
      for (let i = 0; i < 100; i++) {
        expect(i).toBeGreaterThanOrEqual(0)
        expect(i).toBeLessThan(100)
        expect(typeof i).toBe('number')
      }
      
      const end = performance.now()
      const duration = end - start
      
      expect(duration).toBeLessThan(10) // Should be very fast
    })
  })
})