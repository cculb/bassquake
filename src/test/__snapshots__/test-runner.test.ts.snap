// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Runner Validation > Snapshot Testing Support > can create and match snapshots 1`] = `
{
  "features": [
    "sequencer",
    "synthesizer",
    "motion-control",
  ],
  "name": "Bassquake",
  "version": "1.0.0",
}
`;
